drivers/modem/modem_cellular.c

```diff
@@ -0,0 +1,1322 @@
+/*
+ * Copyright (c) 2023 Bjarki Arge Andreasen
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/modem/chat.h>
+#include <zephyr/modem/cmux.h>
+#include <zephyr/modem/pipe.h>
+#include <zephyr/modem/ppp.h>
+#include <zephyr/modem/backend/uart.h>
+#include <zephyr/net/ppp.h>
+#include <zephyr/pm/device.h>
+#include <zephyr/sys/atomic.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(modem_cellular, CONFIG_MODEM_LOG_LEVEL);
+
+#include <string.h>
+#include <stdlib.h>
+
+#define MODEM_CELLULAR_POWER_GPIO_PULSE K_MSEC(1500)
+#define MODEM_CELLULAR_RESET_GPIO_PULSE K_MSEC(100)
+#define MODEM_CELLULAR_STARTUP_TIME     K_MSEC(10000)
+#define MODEM_CELLULAR_SHUTDOWN_TIME    K_MSEC(10000)
+
+enum modem_cellular_state {
+	MODEM_CELLULAR_STATE_IDLE = 0,
+	MODEM_CELLULAR_STATE_RESET_PULSE,
+	MODEM_CELLULAR_STATE_POWER_ON_PULSE,
+	MODEM_CELLULAR_STATE_AWAIT_POWER_ON,
+	MODEM_CELLULAR_STATE_RUN_INIT_SCRIPT,
+	MODEM_CELLULAR_STATE_CONNECT_CMUX,
+	MODEM_CELLULAR_STATE_OPEN_DLCI1,
+	MODEM_CELLULAR_STATE_OPEN_DLCI2,
+	MODEM_CELLULAR_STATE_AWAIT_REGISTERED,
+	MODEM_CELLULAR_STATE_RUN_DIAL_SCRIPT,
```

Comment:
The order of these states doesn't make sense. We want to run the dial script to configure and get on the network. Then we should wait for registered and move to the carrier on state. See my 23-07-27 patch for the state changes.

Reply:
The order has been reversed, now the dial script is performed before waiting for the registered event.